money-core/src/main/scala/com/comcast/money/core/Formatters.scala

```diff
@@ -33,14 +48,25 @@ object Formatters {
     new SpanId(traceId, parentId.toLong, selfId.toLong)
   }
 
-  def toHttpHeader(spanId: SpanId): String =
-    HttpHeaderFormat.format(spanId.traceId, spanId.parentId, spanId.selfId)
+  def toHttpHeader(spanId: SpanId): String = HttpHeaderFormat.format(spanId.traceId, spanId.parentId, spanId.selfId)
 
   def fromB3HttpHeaders(traceId: String, maybeParentSpanId: Option[String], maybeSpanId: Option[String]) = Try {
     (maybeParentSpanId, maybeSpanId) match {
-      case (Some(ps), Some(s)) => new SpanId(traceId, ps.toLong, s.toLong)
-      case (Some(ps), _) => new SpanId(traceId, ps.toLong)
-      case _ => new SpanId(traceId)
+      case (Some(ps), Some(s)) => new SpanId(traceId.toGuid, ps.fromHexStringToLong, s.fromHexStringToLong)
+      case (Some(ps), _) => new SpanId(traceId.toGuid, ps.fromHexStringToLong)
+      case (_, Some(s)) => new SpanId(traceId.toGuid, s.fromHexStringToLong, s.fromHexStringToLong) // root span
+      case _ => new SpanId(traceId.toGuid)
+    }
+  }
+
+  def toB3Headers(spanId: SpanId)(ft: String => Unit, fps: String => Unit, fs: String => Unit): Unit = {
```

Comment:
I'm not sure I understand the reasoning behind three separate methods. I would prefer the below

```
def toB3Headers(spanId: SpanId, addHeader: (String, String) => Unit): Unit = {
    val formatGuid = {
      // X-B3 style traceId's can be 8 octets long
      val traceIdHex = spanId.traceId.fromGuid
      if (traceIdHex.endsWith("0" * 16)) traceIdHex.substring(0, 16)
      else traceIdHex
    }
    addHeader("X-B3-TraceId", formatGuid)
    // No X-b3 parent if this is a root span
    if (spanId.parentId != spanId.selfId) addHeader("X-B3-ParentSpanId", spanId.parentId.toHexString)
    addHeader("X-B3-SpanId", spanId.selfId.toHexString)
  }
```